woocommerce/woocommerce-gateway-stripe
# SCA Subscriptions: Reminder Emails

Fixes the e-mail part of #937 .

#### Changes proposed in this Pull Request:

This PR handles email notifications, related to off-session payments, which require payment authentication. Both Subscriptions and Pre-Orders are handled.

Both extensions would send an email upon a failed payment. This PR replaces the built-in emails with similar ones, which use different wording and subjects, in order to make it clear that instead of a failed payment, the current state requires authentication.

- `WC_Stripe_Email_Failed_Authentication` is a base class, which mainly exists to avoid DRY in the context of using email templates from the extension.
- `WC_Stripe_Email_Failed_Renewal_Authentication` handles WooCommerce Subscriptions renewals.
- `WC_Stripe_Email_Failed_Preorder_Authentication` handles Pre-Orders charges.

Both emails are only sent when particular conditions are met, and before they are sent, the original email, which would be sent from the respective extension, is disconnected.

The `wc_gateway_stripe_process_payment_authentication_required` action is used to trigger those emails.

#### Testing instructions

0. Make sure that you can see (all) emails, sent from the server. I recommend using a free service like MailTrap for this.
1. Create a subscription with a free trial. This way the card will not be set up for future payments.
2. Sign up for the subscription with an SCA card (ex. `4000002500003155`).
3. In the admin, navigate to the new subscription.
4. In the "Subscription Actions" meta box, select "Process renewal".

You should have received an email, looking like this:

<img width="630" alt="image" src="https://user-images.githubusercontent.com/5311119/65917821-8f696180-e3d8-11e9-9d25-705ac5be1535.png">

For pre-orders the steps are pretty much the same, the last one (4) requires you to go to WooCommerce > Pre-Orders -> Actions (Tab) -> Complete (Tab), and submit the form there.


-------------------
